x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/retry/RetryingHttpSender.java

```diff
@@ -0,0 +1,152 @@
+
+                attemptRetry(
+                    request,
+                    responseHandler,
+                    new ElasticsearchException(format("Failed to process request [%s]", request.getRequestLine()), e),
+                    failureCount,
+                    listener
+                );
+            }
+        }, e -> {
+            logException(request, responseHandler.getRequestType(), e);
+            attemptRetry(request, responseHandler, e, failureCount, listener);
+        });
+
+        sender.send(request, responseListener);
+    }
+
+    private void logException(HttpRequestBase request, String requestType, Exception exception) {
+        throttlerManager.warn(
+            logger,
+            format("Failed while sending request [%s] of type [%s]", request.getRequestLine(), requestType),
+            exception
+        );
+    }
+
+    private void logException(HttpRequestBase request, HttpResult result, String requestType, IOException exception) {
+        throttlerManager.warn(
+            logger,
+            format(
+                "Failed to parse the response for request [%s] of type [%s] with status [%s] [%s]",
+                request.getRequestLine(),
+                requestType,
+                result.response().getStatusLine().getStatusCode(),
+                result.response().getStatusLine().getReasonPhrase()
+            ),
+            exception
+        );
+    }
+
+    private void logException(HttpRequestBase request, HttpResult result, String requestType, Exception exception) {
+        throttlerManager.warn(
+            logger,
+            format(
+                "Failed to process the response for request [%s] of type [%s] with status [%s] [%s]",
+                request.getRequestLine(),
+                requestType,
+                result.response().getStatusLine().getStatusCode(),
+                result.response().getStatusLine().getReasonPhrase()
+            ),
+            exception
+        );
+    }
+
+    private void attemptRetry(
+        HttpRequestBase request,
+        ResponseHandler responseHandler,
+        Exception exception,
+        int failureCount,
+        ActionListener<InferenceResults> listener
+    ) {
+        try {
+            // Let's retry all exceptions for now
+            if (failureCount >= retrySettings.getMaxRetries()) {
+                listener.onFailure(
+                    new ElasticsearchException(
+                        format("Failed sending request [%s] after [%s] retries", request.getRequestLine(), retrySettings.getMaxRetries()),
+                        exception
+                    )
+                );
+                return;
+            }
+
+            classLogger.debug(
+                () -> format("Failed sending request [%s], [%s] times, retrying", request.getRequestLine(), failureCount),
+                exception
+            );
+
+            TimeUnit.SECONDS.sleep(retrySettings.getRetryWaitTime().getSeconds());
```

Comment:
This should use an `Executor` and schedule the retry rather than sleeping. See RetryableAction for inspiration https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/support/RetryableAction.java

Reply:
Sounds good. Thanks for the link, didn't even realize we had that! I refactored this class so it leverages that action directly. So we get exponential backoff for free now 🙌